File: transformometro-api/tm_app/domain/diagram/bpmn_node_catalog.py
"""Catálogo canônico de tipos BPMN suportados pelo flowchart_v1."""

from __future__ import annotations

from typing import Any, TypedDict


class BpmnNodeSpec(TypedDict, total=False):
    bpmn_tag: str
    bpmn_event_definition: str | None
    participates_in_flow: bool
    shape: str
    manual_task: bool


BPMN_NODE_CATALOG: dict[str, BpmnNodeSpec] = {
    "start": {"bpmn_tag": "startEvent", "participates_in_flow": True, "shape": "event_start"},
    "start_message": {
        "bpmn_tag": "startEvent",
        "bpmn_event_definition": "messageEventDefinition",
        "participates_in_flow": True,
        "shape": "event_start",
    },
    "start_timer": {
        "bpmn_tag": "startEvent",
        "bpmn_event_definition": "timerEventDefinition",
        "participates_in_flow": True,
        "shape": "event_start",
    },
    "start_signal": {
        "bpmn_tag": "startEvent",
        "bpmn_event_definition": "signalEventDefinition",
        "participates_in_flow": True,
        "shape": "event_start",
    },
    "start_conditional": {
        "bpmn_tag": "startEvent",
        "bpmn_event_definition": "conditionalEventDefinition",
        "participates_in_flow": True,
        "shape": "event_start",
    },
    "start_multiple": {
        "bpmn_tag": "startEvent",
        "bpmn_event_definition": "multipleEventDefinition",
        "participates_in_flow": True,
        "shape": "event_start",
    },
    "start_parallel": {
        "bpmn_tag": "startEvent",
        "bpmn_event_definition": "parallelMultipleEventDefinition",
        "participates_in_flow": True,
        "shape": "event_start",
    },
    "intermediate": {
        "bpmn_tag": "intermediateCatchEvent",
        "participates_in_flow": True,
        "shape": "event_intermediate_catch",
    },
    "intermediate_message_catch": {
        "bpmn_tag": "intermediateCatchEvent",
        "bpmn_event_definition": "messageEventDefinition",
        "participates_in_flow": True,
        "shape": "event_intermediate_catch",
    },
    "intermediate_timer": {
        "bpmn_tag": "intermediateCatchEvent",
        "bpmn_event_definition": "timerEventDefinition",
        "participates_in_flow": True,
        "shape": "event_intermediate_catch",
    },
    "intermediate_signal_catch": {
        "bpmn_tag": "intermediateCatchEvent",
        "bpmn_event_definition": "signalEventDefinition",
        "participates_in_flow": True,
        "shape": "event_intermediate_catch",
    },
    "intermediate_conditional": {
        "bpmn_tag": "intermediateCatchEvent",
        "bpmn_event_definition": "conditionalEventDefinition",
        "participates_in_flow": True,
        "shape": "event_intermediate_catch",
    },
    "intermediate_link_catch": {
        "bpmn_tag": "intermediateCatchEvent",
        "bpmn_event_definition": "linkEventDefinition",
        "participates_in_flow": True,
        "shape": "event_intermediate_catch",
    },
    "intermediate_message_throw": {
        "bpmn_tag": "intermediateThrowEvent",
        "bpmn_event_definition": "messageEventDefinition",
        "participates_in_flow": True,
        "shape": "event_intermediate_throw",
    },
    "intermediate_signal_throw": {
        "bpmn_tag": "intermediateThrowEvent",
        "bpmn_event_definition": "signalEventDefinition",
        "participates_in_flow": True,
        "shape": "event_intermediate_throw",
    },
    "intermediate_link_throw": {
        "bpmn_tag": "intermediateThrowEvent",
        "bpmn_event_definition": "linkEventDefinition",
        "participates_in_flow": True,
        "shape": "event_intermediate_throw",
    },
    "intermediate_escalation_throw": {
        "bpmn_tag": "intermediateThrowEvent",
        "bpmn_event_definition": "escalationEventDefinition",
        "participates_in_flow": True,
        "shape": "event_intermediate_throw",
    },
    "intermediate_compensation_throw": {
        "bpmn_tag": "intermediateThrowEvent",
        "bpmn_event_definition": "compensateEventDefinition",
        "participates_in_flow": True,
        "shape": "event_intermediate_throw",
    },
    "end": {"bpmn_tag": "endEvent", "participates_in_flow": True, "shape": "event_end"},
    "end_message": {
        "bpmn_tag": "endEvent",
        "bpmn_event_definition": "messageEventDefinition",
        "participates_in_flow": True,
        "shape": "event_end",
    },
    "end_error": {
        "bpmn_tag": "endEvent",
        "bpmn_event_definition": "errorEventDefinition",
        "participates_in_flow": True,
        "shape": "event_end",
    },
    "end_terminate": {
        "bpmn_tag": "endEvent",
        "bpmn_event_definition": "terminateEventDefinition",
        "participates_in_flow": True,
        "shape": "event_end",
    },
    "end_signal": {
        "bpmn_tag": "endEvent",
        "bpmn_event_definition": "signalEventDefinition",
        "participates_in_flow": True,
        "shape": "event_end",
    },
    "end_escalation": {
        "bpmn_tag": "endEvent",
        "bpmn_event_definition": "escalationEventDefinition",
        "participates_in_flow": True,
        "shape": "event_end",
    },
    "end_cancel": {
        "bpmn_tag": "endEvent",
        "bpmn_event_definition": "cancelEventDefinition",
        "participates_in_flow": True,
        "shape": "event_end",
    },
    "end_compensation": {
        "bpmn_tag": "endEvent",
        "bpmn_event_definition": "compensateEventDefinition",
        "participates_in_flow": True,
        "shape": "event_end",
    },
    "decision": {"bpmn_tag": "exclusiveGateway", "participates_in_flow": True, "shape": "gateway"},
    "gateway_parallel": {"bpmn_tag": "parallelGateway", "participates_in_flow": True, "shape": "gateway"},
    "gateway_inclusive": {"bpmn_tag": "inclusiveGateway", "participates_in_flow": True, "shape": "gateway"},
    "gateway_complex": {"bpmn_tag": "complexGateway", "participates_in_flow": True, "shape": "gateway"},
    "gateway_event": {"bpmn_tag": "eventBasedGateway", "participates_in_flow": True, "shape": "gateway"},
    "process": {"bpmn_tag": "task", "participates_in_flow": True, "shape": "task", "manual_task": True},
    "task_user": {"bpmn_tag": "userTask", "participates_in_flow": True, "shape": "task"},
    "task_service": {"bpmn_tag": "serviceTask", "participates_in_flow": True, "shape": "task"},
    "task_manual": {"bpmn_tag": "manualTask", "participates_in_flow": True, "shape": "task", "manual_task": True},
    "task_script": {"bpmn_tag": "scriptTask", "participates_in_flow": True, "shape": "task"},
    "task_business_rule": {"bpmn_tag": "businessRuleTask", "participates_in_flow": True, "shape": "task"},
    "task_send": {"bpmn_tag": "sendTask", "participates_in_flow": True, "shape": "task"},
    "task_receive": {"bpmn_tag": "receiveTask", "participates_in_flow": True, "shape": "task"},
    "subprocess": {"bpmn_tag": "subProcess", "participates_in_flow": True, "shape": "activity_subprocess"},
    "call_activity": {"bpmn_tag": "callActivity", "participates_in_flow": True, "shape": "activity_call"},
    "subprocess_ad_hoc": {"bpmn_tag": "adHocSubProcess", "participates_in_flow": True, "shape": "activity_ad_hoc"},
    "subprocess_transaction": {"bpmn_tag": "transaction", "participates_in_flow": True, "shape": "activity_transaction"},
    "subprocess_event": {"bpmn_tag": "subProcess", "participates_in_flow": True, "shape": "activity_event_subprocess"},
    "document": {"bpmn_tag": "textAnnotation", "participates_in_flow": False, "shape": "artifact_document"},
    "data_object": {"bpmn_tag": "dataObjectReference", "participates_in_flow": False, "shape": "artifact_data_object"},
    "data": {"bpmn_tag": "dataStoreReference", "participates_in_flow": False, "shape": "artifact_data_store"},
    "comment": {"bpmn_tag": "textAnnotation", "participates_in_flow": False, "shape": "artifact_comment"},
    "group": {"bpmn_tag": "group", "participates_in_flow": False, "shape": "artifact_group"},
    "boundary_timer": {
        "bpmn_tag": "boundaryEvent",
        "bpmn_event_definition": "timerEventDefinition",
        "participates_in_flow": True,
        "shape": "boundary",
    },
    "boundary_message": {
        "bpmn_tag": "boundaryEvent",
        "bpmn_event_definition": "messageEventDefinition",
        "participates_in_flow": True,
        "shape": "boundary",
    },
    "boundary_error": {
        "bpmn_tag": "boundaryEvent",
        "bpmn_event_definition": "errorEventDefinition",
        "participates_in_flow": True,
        "shape": "boundary",
    },
    "boundary_signal": {
        "bpmn_tag": "boundaryEvent",
        "bpmn_event_definition": "signalEventDefinition",
        "participates_in_flow": True,
        "shape": "boundary",
    },
    "boundary_escalation": {
        "bpmn_tag": "boundaryEvent",
        "bpmn_event_definition": "escalationEventDefinition",
        "participates_in_flow": True,
        "shape": "boundary",
    },
    "boundary_compensation": {
        "bpmn_tag": "boundaryEvent",
        "bpmn_event_definition": "compensateEventDefinition",
        "participates_in_flow": True,
        "shape": "boundary",
    },
    "boundary_cancel": {
        "bpmn_tag": "boundaryEvent",
        "bpmn_event_definition": "cancelEventDefinition",
        "participates_in_flow": True,
        "shape": "boundary",
    },
    "boundary_conditional": {
        "bpmn_tag": "boundaryEvent",
        "bpmn_event_definition": "conditionalEventDefinition",
        "participates_in_flow": True,
        "shape": "boundary",
    },
}

NODE_TYPES = frozenset(BPMN_NODE_CATALOG.keys())

EDGE_KINDS = frozenset({"sequence", "message_flow", "association"})

START_EVENT_TYPES = frozenset(
    node_type for node_type, spec in BPMN_NODE_CATALOG.items() if spec.get("shape") == "event_start"
)
END_EVENT_TYPES = frozenset(
    node_type for node_type, spec in BPMN_NODE_CATALOG.items() if spec.get("shape") == "event_end"
)
GATEWAY_TYPES = frozenset(
    node_type for node_type, spec in BPMN_NODE_CATALOG.items() if spec.get("shape") == "gateway"
)
NON_FLOW_NODE_TYPES = frozenset(
    node_type
    for node_type, spec in BPMN_NODE_CATALOG.items()
    if not spec.get("participates_in_flow", True)
)

BPMN_TAG_TO_DEFAULT_TYPE: dict[str, str] = {
    "startEvent": "start",
    "endEvent": "end",
    "intermediateCatchEvent": "intermediate",
    "intermediateThrowEvent": "intermediate_message_throw",
    "boundaryEvent": "boundary_timer",
    "exclusiveGateway": "decision",
    "parallelGateway": "gateway_parallel",
    "inclusiveGateway": "gateway_inclusive",
    "complexGateway": "gateway_complex",
    "eventBasedGateway": "gateway_event",
    "task": "process",
    "userTask": "task_user",
    "serviceTask": "task_service",
    "manualTask": "task_manual",
    "scriptTask": "task_script",
    "businessRuleTask": "task_business_rule",
    "sendTask": "task_send",
    "receiveTask": "task_receive",
    "subProcess": "subprocess",
    "callActivity": "call_activity",
    "adHocSubProcess": "subprocess_ad_hoc",
    "transaction": "subprocess_transaction",
    "textAnnotation": "comment",
    "dataObjectReference": "data_object",
    "dataStoreReference": "data",
    "group": "group",
}

EVENT_DEFINITION_SUFFIX: dict[str, str] = {
    "messageEventDefinition": "message",
    "timerEventDefinition": "timer",
    "signalEventDefinition": "signal",
    "conditionalEventDefinition": "conditional",
    "errorEventDefinition": "error",
    "escalationEventDefinition": "escalation",
    "compensateEventDefinition": "compensation",
    "cancelEventDefinition": "cancel",
    "terminateEventDefinition": "terminate",
    "linkEventDefinition": "link",
    "multipleEventDefinition": "multiple",
    "parallelMultipleEventDefinition": "parallel",
}


def catalog_spec(node_type: str) -> BpmnNodeSpec:
    return BPMN_NODE_CATALOG.get(node_type, BPMN_NODE_CATALOG["process"])


def normalize_node_type(node_type: str) -> str:
    return node_type if node_type in NODE_TYPES else "process"

File: plugins/plugin-ui/src/components/bpmn/model/bpmnNodeCatalog.ts
export type BpmnPaletteCategoryId =
  | "events_start"
  | "events_intermediate"
  | "events_end"
  | "gateways"
  | "tasks"
  | "activities"
  | "artifacts"
  | "boundary";

export type BpmnShapeFamily =
  | "event_start"
  | "event_end"
  | "event_intermediate_catch"
  | "event_intermediate_throw"
  | "gateway"
  | "task"
  | "activity_subprocess"
  | "activity_call"
  | "activity_ad_hoc"
  | "activity_transaction"
  | "activity_event_subprocess"
  | "artifact_document"
  | "artifact_data_store"
  | "artifact_data_object"
  | "artifact_comment"
  | "artifact_group"
  | "boundary";

export type BpmnMarker =
  | "none"
  | "message"
  | "timer"
  | "signal"
  | "conditional"
  | "error"
  | "escalation"
  | "compensation"
  | "cancel"
  | "link"
  | "multiple"
  | "parallel"
  | "terminate"
  | "exclusive"
  | "parallel_gateway"
  | "inclusive"
  | "complex"
  | "event_based"
  | "user"
  | "service"
  | "manual"
  | "script"
  | "business_rule"
  | "send"
  | "receive"
  | "call"
  | "ad_hoc"
  | "transaction"
  | "event_sub";

export type FlowchartNodeType = keyof typeof BPMN_NODE_DEFINITIONS;

export type BpmnNodeDefinition = {
  label: string;
  category: BpmnPaletteCategoryId;
  shape: BpmnShapeFamily;
  marker: BpmnMarker;
  bpmnTag: string;
  bpmnEventDefinition?: string;
  participatesInFlow: boolean;
  hint: string;
  manualTask?: boolean;
};

export const BPMN_PALETTE_CATEGORIES: Array<{ id: BpmnPaletteCategoryId; label: string }> = [
  { id: "events_start", label: "Eventos — início" },
  { id: "events_intermediate", label: "Eventos — intermediários" },
  { id: "events_end", label: "Eventos — fim" },
  { id: "gateways", label: "Desvios" },
  { id: "tasks", label: "Tarefas" },
  { id: "activities", label: "Atividades" },
  { id: "artifacts", label: "Artefatos" },
  { id: "boundary", label: "Eventos de borda" },
];

export const BPMN_NODE_DEFINITIONS = {
  start: {
    label: "Início",
    category: "events_start",
    shape: "event_start",
    marker: "none",
    bpmnTag: "startEvent",
    participatesInFlow: true,
    hint: "Evento de início genérico (nenhum gatilho específico).",
  },
  start_message: {
    label: "Início — mensagem",
    category: "events_start",
    shape: "event_start",
    marker: "message",
    bpmnTag: "startEvent",
    bpmnEventDefinition: "messageEventDefinition",
    participatesInFlow: true,
    hint: "Processo inicia ao receber uma mensagem.",
  },
  start_timer: {
    label: "Início — temporizador",
    category: "events_start",
    shape: "event_start",
    marker: "timer",
    bpmnTag: "startEvent",
    bpmnEventDefinition: "timerEventDefinition",
    participatesInFlow: true,
    hint: "Processo inicia em data/horário ou intervalo programado.",
  },
  start_signal: {
    label: "Início — sinal",
    category: "events_start",
    shape: "event_start",
    marker: "signal",
    bpmnTag: "startEvent",
    bpmnEventDefinition: "signalEventDefinition",
    participatesInFlow: true,
    hint: "Processo inicia ao receber um sinal em difusão.",
  },
  start_conditional: {
    label: "Início — condicional",
    category: "events_start",
    shape: "event_start",
    marker: "conditional",
    bpmnTag: "startEvent",
    bpmnEventDefinition: "conditionalEventDefinition",
    participatesInFlow: true,
    hint: "Processo inicia quando uma condição de negócio é satisfeita.",
  },
  start_multiple: {
    label: "Início — múltiplo",
    category: "events_start",
    shape: "event_start",
    marker: "multiple",
    bpmnTag: "startEvent",
    bpmnEventDefinition: "multipleEventDefinition",
    participatesInFlow: true,
    hint: "Início disparado por um entre vários eventos possíveis.",
  },
  start_parallel: {
    label: "Início — paralelo",
    category: "events_start",
    shape: "event_start",
    marker: "parallel",
    bpmnTag: "startEvent",
    bpmnEventDefinition: "parallelMultipleEventDefinition",
    participatesInFlow: true,
    hint: "Início exige que todos os eventos paralelos ocorram.",
  },

  intermediate: {
    label: "Intermediário",
    category: "events_intermediate",
    shape: "event_intermediate_catch",
    marker: "none",
    bpmnTag: "intermediateCatchEvent",
    participatesInFlow: true,
    hint: "Evento intermediário genérico (captura).",
  },
  intermediate_message_catch: {
    label: "Intermediário — mensagem (recepção)",
    category: "events_intermediate",
    shape: "event_intermediate_catch",
    marker: "message",
    bpmnTag: "intermediateCatchEvent",
    bpmnEventDefinition: "messageEventDefinition",
    participatesInFlow: true,
    hint: "Aguarda recebimento de mensagem durante o fluxo.",
  },
  intermediate_timer: {
    label: "Intermediário — temporizador",
    category: "events_intermediate",
    shape: "event_intermediate_catch",
    marker: "timer",
    bpmnTag: "intermediateCatchEvent",
    bpmnEventDefinition: "timerEventDefinition",
    participatesInFlow: true,
    hint: "Aguarda tempo decorrer (delay, prazo, ciclo).",
  },
  intermediate_signal_catch: {
    label: "Intermediário — sinal (recepção)",
    category: "events_intermediate",
    shape: "event_intermediate_catch",
    marker: "signal",
    bpmnTag: "intermediateCatchEvent",
    bpmnEventDefinition: "signalEventDefinition",
    participatesInFlow: true,
    hint: "Aguarda um sinal em difusão durante o fluxo.",
  },
  intermediate_conditional: {
    label: "Intermediário — condicional",
    category: "events_intermediate",
    shape: "event_intermediate_catch",
    marker: "conditional",
    bpmnTag: "intermediateCatchEvent",
    bpmnEventDefinition: "conditionalEventDefinition",
    participatesInFlow: true,
    hint: "Aguarda condição de negócio ser satisfeita.",
  },
  intermediate_link_catch: {
    label: "Intermediário — vínculo (recepção)",
    category: "events_intermediate",
    shape: "event_intermediate_catch",
    marker: "link",
    bpmnTag: "intermediateCatchEvent",
    bpmnEventDefinition: "linkEventDefinition",
    participatesInFlow: true,
    hint: "Ponto de chegada de um vínculo dentro do mesmo processo.",
  },
  intermediate_message_throw: {
    label: "Intermediário — mensagem (envio)",
    category: "events_intermediate",
    shape: "event_intermediate_throw",
    marker: "message",
    bpmnTag: "intermediateThrowEvent",
    bpmnEventDefinition: "messageEventDefinition",
    participatesInFlow: true,
    hint: "Envia mensagem para outro participante ou processo.",
  },
  intermediate_signal_throw: {
    label: "Intermediário — sinal (envio)",
    category: "events_intermediate",
    shape: "event_intermediate_throw",
    marker: "signal",
    bpmnTag: "intermediateThrowEvent",
    bpmnEventDefinition: "signalEventDefinition",
    participatesInFlow: true,
    hint: "Emite sinal em difusão durante o fluxo.",
  },
  intermediate_link_throw: {
    label: "Intermediário — vínculo (envio)",
    category: "events_intermediate",
    shape: "event_intermediate_throw",
    marker: "link",
    bpmnTag: "intermediateThrowEvent",
    bpmnEventDefinition: "linkEventDefinition",
    participatesInFlow: true,
    hint: "Salta para o ponto de vínculo correspondente no diagrama.",
  },
  intermediate_escalation_throw: {
    label: "Intermediário — escalação",
    category: "events_intermediate",
    shape: "event_intermediate_throw",
    marker: "escalation",
    bpmnTag: "intermediateThrowEvent",
    bpmnEventDefinition: "escalationEventDefinition",
    participatesInFlow: true,
    hint: "Dispara escalação para nível superior ou subprocesso.",
  },
  intermediate_compensation_throw: {
    label: "Intermediário — compensação",
    category: "events_intermediate",
    shape: "event_intermediate_throw",
    marker: "compensation",
    bpmnTag: "intermediateThrowEvent",
    bpmnEventDefinition: "compensateEventDefinition",
    participatesInFlow: true,
    hint: "Solicita compensação de atividades já concluídas.",
  },

  end: {
    label: "Fim",
    category: "events_end",
    shape: "event_end",
    marker: "none",
    bpmnTag: "endEvent",
    participatesInFlow: true,
    hint: "Evento de fim genérico — término normal do fluxo.",
  },
  end_message: {
    label: "Fim — mensagem",
    category: "events_end",
    shape: "event_end",
    marker: "message",
    bpmnTag: "endEvent",
    bpmnEventDefinition: "messageEventDefinition",
    participatesInFlow: true,
    hint: "Fim ao enviar mensagem para outro participante.",
  },
  end_error: {
    label: "Fim — erro",
    category: "events_end",
    shape: "event_end",
    marker: "error",
    bpmnTag: "endEvent",
    bpmnEventDefinition: "errorEventDefinition",
    participatesInFlow: true,
    hint: "Fim por erro de negócio ou técnico.",
  },
  end_terminate: {
    label: "Fim — terminate",
    category: "events_end",
    shape: "event_end",
    marker: "terminate",
    bpmnTag: "endEvent",
    bpmnEventDefinition: "terminateEventDefinition",
    participatesInFlow: true,
    hint: "Encerra imediatamente todo o processo (inclui tokens paralelos).",
  },
  end_signal: {
    label: "Fim — sinal",
    category: "events_end",
    shape: "event_end",
    marker: "signal",
    bpmnTag: "endEvent",
    bpmnEventDefinition: "signalEventDefinition",
    participatesInFlow: true,
    hint: "Fim ao emitir sinal em difusão.",
  },
  end_escalation: {
    label: "Fim — escalação",
    category: "events_end",
    shape: "event_end",
    marker: "escalation",
    bpmnTag: "endEvent",
    bpmnEventDefinition: "escalationEventDefinition",
    participatesInFlow: true,
    hint: "Fim por escalação para responsável externo.",
  },
  end_cancel: {
    label: "Fim — cancelamento",
    category: "events_end",
    shape: "event_end",
    marker: "cancel",
    bpmnTag: "endEvent",
    bpmnEventDefinition: "cancelEventDefinition",
    participatesInFlow: true,
    hint: "Fim por cancelamento de transação.",
  },
  end_compensation: {
    label: "Fim — compensação",
    category: "events_end",
    shape: "event_end",
    marker: "compensation",
    bpmnTag: "endEvent",
    bpmnEventDefinition: "compensateEventDefinition",
    participatesInFlow: true,
    hint: "Fim após concluir fluxo de compensação.",
  },

  decision: {
    label: "Decisão (XOR)",
    category: "gateways",
    shape: "gateway",
    marker: "exclusive",
    bpmnTag: "exclusiveGateway",
    participatesInFlow: true,
    hint: "Desvio exclusivo — apenas um caminho de saída é tomado.",
  },
  gateway_parallel: {
    label: "Paralelo (AND)",
    category: "gateways",
    shape: "gateway",
    marker: "parallel_gateway",
    bpmnTag: "parallelGateway",
    participatesInFlow: true,
    hint: "Desvio paralelo — todos os caminhos de saída são ativados.",
  },
  gateway_inclusive: {
    label: "Inclusivo (OR)",
    category: "gateways",
    shape: "gateway",
    marker: "inclusive",
    bpmnTag: "inclusiveGateway",
    participatesInFlow: true,
    hint: "Desvio inclusivo — um ou mais caminhos conforme condição.",
  },
  gateway_complex: {
    label: "Complexo",
    category: "gateways",
    shape: "gateway",
    marker: "complex",
    bpmnTag: "complexGateway",
    participatesInFlow: true,
    hint: "Desvio com regras de sincronização customizadas.",
  },
  gateway_event: {
    label: "Baseado em evento",
    category: "gateways",
    shape: "gateway",
    marker: "event_based",
    bpmnTag: "eventBasedGateway",
    participatesInFlow: true,
    hint: "Escolhe caminho conforme o primeiro evento que ocorrer.",
  },

  process: {
    label: "Atividade",
    category: "tasks",
    shape: "task",
    marker: "none",
    bpmnTag: "task",
    participatesInFlow: true,
    manualTask: true,
    hint: "Tarefa genérica ou manual (retângulo arredondado).",
  },
  task_user: {
    label: "Tarefa — usuário",
    category: "tasks",
    shape: "task",
    marker: "user",
    bpmnTag: "userTask",
    participatesInFlow: true,
    hint: "Tarefa executada por um usuário humano.",
  },
  task_service: {
    label: "Tarefa — serviço",
    category: "tasks",
    shape: "task",
    marker: "service",
    bpmnTag: "serviceTask",
    participatesInFlow: true,
    hint: "Tarefa automatizada via serviço ou sistema.",
  },
  task_manual: {
    label: "Tarefa — manual",
    category: "tasks",
    shape: "task",
    marker: "manual",
    bpmnTag: "manualTask",
    participatesInFlow: true,
    manualTask: true,
    hint: "Trabalho manual sem suporte de sistema.",
  },
  task_script: {
    label: "Tarefa — script",
    category: "tasks",
    shape: "task",
    marker: "script",
    bpmnTag: "scriptTask",
    participatesInFlow: true,
    hint: "Tarefa executada por script embarcado.",
  },
  task_business_rule: {
    label: "Tarefa — regra de negócio",
    category: "tasks",
    shape: "task",
    marker: "business_rule",
    bpmnTag: "businessRuleTask",
    participatesInFlow: true,
    hint: "Decisão automatizada via motor de regras (DMN).",
  },
  task_send: {
    label: "Tarefa — enviar",
    category: "tasks",
    shape: "task",
    marker: "send",
    bpmnTag: "sendTask",
    participatesInFlow: true,
    hint: "Envia mensagem como parte da atividade.",
  },
  task_receive: {
    label: "Tarefa — receber",
    category: "tasks",
    shape: "task",
    marker: "receive",
    bpmnTag: "receiveTask",
    participatesInFlow: true,
    hint: "Aguarda e recebe mensagem como atividade.",
  },

  subprocess: {
    label: "Subprocesso",
    category: "activities",
    shape: "activity_subprocess",
    marker: "none",
    bpmnTag: "subProcess",
    participatesInFlow: true,
    hint: "Subprocesso encapsulado com fluxo interno.",
  },
  call_activity: {
    label: "Chamada de subprocesso",
    category: "activities",
    shape: "activity_call",
    marker: "call",
    bpmnTag: "callActivity",
    participatesInFlow: true,
    hint: "Referência reutilizável a outro diagrama de processo.",
  },
  subprocess_ad_hoc: {
    label: "Subprocesso improvisado",
    category: "activities",
    shape: "activity_ad_hoc",
    marker: "ad_hoc",
    bpmnTag: "adHocSubProcess",
    participatesInFlow: true,
    hint: "Subprocesso não ordenado — atividades executadas conforme necessidade.",
  },
  subprocess_transaction: {
    label: "Subprocesso transação",
    category: "activities",
    shape: "activity_transaction",
    marker: "transaction",
    bpmnTag: "transaction",
    participatesInFlow: true,
    hint: "Subprocesso transacional com compensação/cancelamento.",
  },
  subprocess_event: {
    label: "Subprocesso por evento",
    category: "activities",
    shape: "activity_event_subprocess",
    marker: "event_sub",
    bpmnTag: "subProcess",
    bpmnEventDefinition: "triggeredByEvent",
    participatesInFlow: true,
    hint: "Subprocesso acionado por evento.",
  },

  document: {
    label: "Documento",
    category: "artifacts",
    shape: "artifact_document",
    marker: "none",
    bpmnTag: "textAnnotation",
    participatesInFlow: false,
    hint: "Documento ou artefato produzido/consumido (folha com canto dobrado).",
  },
  data_object: {
    label: "Objeto de dados",
    category: "artifacts",
    shape: "artifact_data_object",
    marker: "none",
    bpmnTag: "dataObjectReference",
    participatesInFlow: false,
    hint: "Dado de entrada/saída referenciado no fluxo (paralelogramo).",
  },
  data: {
    label: "Armazenamento",
    category: "artifacts",
    shape: "artifact_data_store",
    marker: "none",
    bpmnTag: "dataStoreReference",
    participatesInFlow: false,
    hint: "Armazenamento persistente (cilindro BPMN).",
  },
  comment: {
    label: "Nota",
    category: "artifacts",
    shape: "artifact_comment",
    marker: "none",
    bpmnTag: "textAnnotation",
    participatesInFlow: false,
    hint: "Anotação explicativa — não participa da lógica do fluxo.",
  },
  group: {
    label: "Grupo",
    category: "artifacts",
    shape: "artifact_group",
    marker: "none",
    bpmnTag: "group",
    participatesInFlow: false,
    hint: "Agrupamento visual de elementos relacionados.",
  },

  boundary_timer: {
    label: "Borda — temporizador",
    category: "boundary",
    shape: "boundary",
    marker: "timer",
    bpmnTag: "boundaryEvent",
    bpmnEventDefinition: "timerEventDefinition",
    participatesInFlow: true,
    hint: "Interrompe ou desvia atividade após prazo.",
  },
  boundary_message: {
    label: "Borda — mensagem",
    category: "boundary",
    shape: "boundary",
    marker: "message",
    bpmnTag: "boundaryEvent",
    bpmnEventDefinition: "messageEventDefinition",
    participatesInFlow: true,
    hint: "Desvia fluxo ao receber mensagem durante atividade.",
  },
  boundary_error: {
    label: "Borda — erro",
    category: "boundary",
    shape: "boundary",
    marker: "error",
    bpmnTag: "boundaryEvent",
    bpmnEventDefinition: "errorEventDefinition",
    participatesInFlow: true,
    hint: "Captura erro da atividade anexada (interrompente).",
  },
  boundary_signal: {
    label: "Borda — sinal",
    category: "boundary",
    shape: "boundary",
    marker: "signal",
    bpmnTag: "boundaryEvent",
    bpmnEventDefinition: "signalEventDefinition",
    participatesInFlow: true,
    hint: "Desvia fluxo ao receber sinal durante atividade.",
  },
  boundary_escalation: {
    label: "Borda — escalação",
    category: "boundary",
    shape: "boundary",
    marker: "escalation",
    bpmnTag: "boundaryEvent",
    bpmnEventDefinition: "escalationEventDefinition",
    participatesInFlow: true,
    hint: "Escala atividade para responsável externo.",
  },
  boundary_compensation: {
    label: "Borda — compensação",
    category: "boundary",
    shape: "boundary",
    marker: "compensation",
    bpmnTag: "boundaryEvent",
    bpmnEventDefinition: "compensateEventDefinition",
    participatesInFlow: true,
    hint: "Gatilho de compensação na atividade anexada.",
  },
  boundary_cancel: {
    label: "Borda — cancelamento",
    category: "boundary",
    shape: "boundary",
    marker: "cancel",
    bpmnTag: "boundaryEvent",
    bpmnEventDefinition: "cancelEventDefinition",
    participatesInFlow: true,
    hint: "Cancela transação da atividade anexada.",
  },
  boundary_conditional: {
    label: "Borda — condicional",
    category: "boundary",
    shape: "boundary",
    marker: "conditional",
    bpmnTag: "boundaryEvent",
    bpmnEventDefinition: "conditionalEventDefinition",
    participatesInFlow: true,
    hint: "Desvia fluxo quando condição é satisfeita durante atividade.",
  },
} as const satisfies Record<string, BpmnNodeDefinition>;

export const FLOWCHART_NODE_TYPES = Object.keys(BPMN_NODE_DEFINITIONS) as FlowchartNodeType[];

export const FLOWCHART_NODE_PALETTE = FLOWCHART_NODE_TYPES.map((type) => ({
  type,
  label: BPMN_NODE_DEFINITIONS[type].label,
  category: BPMN_NODE_DEFINITIONS[type].category,
}));

export function getBpmnNodeDefinition(type: string): BpmnNodeDefinition | undefined {
  if (!isKnownFlowchartNodeType(type)) return undefined;
  return BPMN_NODE_DEFINITIONS[type] as BpmnNodeDefinition;
}

export function isManualTaskType(type: string): boolean {
  const def = getBpmnNodeDefinition(type);
  return Boolean(def?.manualTask) || type === "process";
}

export function isKnownFlowchartNodeType(type: string): type is FlowchartNodeType {
  return type in BPMN_NODE_DEFINITIONS;
}

export function normalizeFlowchartNodeType(type: string): FlowchartNodeType {
  return isKnownFlowchartNodeType(type) ? type : "process";
}

const START_TYPES = new Set(
  FLOWCHART_NODE_TYPES.filter((type) => BPMN_NODE_DEFINITIONS[type].shape === "event_start")
);
const END_TYPES = new Set(
  FLOWCHART_NODE_TYPES.filter((type) => BPMN_NODE_DEFINITIONS[type].shape === "event_end")
);
const GATEWAY_TYPES = new Set(
  FLOWCHART_NODE_TYPES.filter((type) => BPMN_NODE_DEFINITIONS[type].shape === "gateway")
);
const NON_FLOW_TYPES = new Set(
  FLOWCHART_NODE_TYPES.filter((type) => !BPMN_NODE_DEFINITIONS[type].participatesInFlow)
);

export function isStartEventType(type: string): boolean {
  return START_TYPES.has(type as FlowchartNodeType);
}

export function isEndEventType(type: string): boolean {
  return END_TYPES.has(type as FlowchartNodeType);
}

export function isGatewayType(type: string): boolean {
  return GATEWAY_TYPES.has(type as FlowchartNodeType);
}

export function isNonFlowNodeType(type: string): boolean {
  return NON_FLOW_TYPES.has(type as FlowchartNodeType);
}

export function paletteByCategory(category: BpmnPaletteCategoryId) {
  return FLOWCHART_NODE_PALETTE.filter((item) => item.category === category);
}
